From 8f50dc03acb2d81daace93188a5fd1098c0251e0 Mon Sep 17 00:00:00 2001
From: andrewshaoyu <734921034@qq.com>
Date: Mon, 27 Dec 2021 09:54:54 +0800
Subject: During SyncAll(), do syncDiscoveryNamespaces before sync
 pods,services,endpoints (#36628)

* During SyncAll(), do syncDiscoveryNamespaces before others

* judge discoverySelectors first
---
 .../kube/controller/controller.go             |  9 +++++
 .../controller/filter/discoverynamespaces.go  | 36 +++++++++++++++++++
 2 files changed, 45 insertions(+)

diff --git a/pilot/pkg/serviceregistry/kube/controller/controller.go b/pilot/pkg/serviceregistry/kube/controller/controller.go
index 9a9ea5f327..cbc479786e 100644
--- a/pilot/pkg/serviceregistry/kube/controller/controller.go
+++ b/pilot/pkg/serviceregistry/kube/controller/controller.go
@@ -781,6 +781,7 @@ func (c *Controller) informersSynced() bool {
 func (c *Controller) SyncAll() error {
 	c.beginSync.Store(true)
 	var err *multierror.Error
+	err = multierror.Append(err, c.syncDiscoveryNamespaces())
 	err = multierror.Append(err, c.syncSystemNamespace())
 	err = multierror.Append(err, c.syncNodes())
 	err = multierror.Append(err, c.syncServices())
@@ -802,6 +803,14 @@ func (c *Controller) syncSystemNamespace() error {
 	return err
 }
 
+func (c *Controller) syncDiscoveryNamespaces() error {
+	var err error
+	if c.nsLister != nil {
+		err = c.opts.DiscoveryNamespacesFilter.SyncNamespaces()
+	}
+	return err
+}
+
 func (c *Controller) syncNodes() error {
 	var err *multierror.Error
 	nodes := c.nodeInformer.GetIndexer().List()
diff --git a/pilot/pkg/serviceregistry/kube/controller/filter/discoverynamespaces.go b/pilot/pkg/serviceregistry/kube/controller/filter/discoverynamespaces.go
index 0664fd1def..fecb3da4db 100644
--- a/pilot/pkg/serviceregistry/kube/controller/filter/discoverynamespaces.go
+++ b/pilot/pkg/serviceregistry/kube/controller/filter/discoverynamespaces.go
@@ -33,6 +33,8 @@ type DiscoveryNamespacesFilter interface {
 	Filter(obj interface{}) bool
 	// SelectorsChanged is invoked when meshConfig's discoverySelectors change, returns any newly selected namespaces and deselected namespaces
 	SelectorsChanged(discoverySelectors []*metav1.LabelSelector) (selectedNamespaces []string, deselectedNamespaces []string)
+	// SyncNamespaces is invoked when namespace informer hasSynced before other controller SyncAll
+	SyncNamespaces() error
 	// NamespaceCreated returns true if the created namespace is selected for discovery
 	NamespaceCreated(ns metav1.ObjectMeta) (membershipChanged bool)
 	// NamespaceUpdated : membershipChanged will be true if the updated namespace is newly selected or deselected for discovery
@@ -142,6 +144,40 @@ func (d *discoveryNamespacesFilter) SelectorsChanged(
 	return
 }
 
+func (d *discoveryNamespacesFilter) SyncNamespaces() error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	newDiscoveryNamespaces := sets.NewString()
+
+	namespaceList, err := d.nsLister.List(labels.Everything())
+	if err != nil {
+		log.Errorf("error initializing discovery namespaces filter, failed to list namespaces: %v", err)
+		return err
+	}
+
+	// omitting discoverySelectors indicates discovering all namespaces
+	if len(d.discoverySelectors) == 0 {
+		for _, ns := range namespaceList {
+			newDiscoveryNamespaces.Insert(ns.Name)
+		}
+	}
+
+	// range over all namespaces to get discovery namespaces
+	for _, ns := range namespaceList {
+		for _, selector := range d.discoverySelectors {
+			if selector.Matches(labels.Set(ns.Labels)) {
+				newDiscoveryNamespaces.Insert(ns.Name)
+			}
+		}
+	}
+
+	// update filter state
+	d.discoveryNamespaces = newDiscoveryNamespaces
+
+	return nil
+}
+
 // NamespaceCreated : if newly created namespace is selected, update namespace membership
 func (d *discoveryNamespacesFilter) NamespaceCreated(ns metav1.ObjectMeta) (membershipChanged bool) {
 	if d.isSelected(ns.Labels) {
-- 
2.35.3

